src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -669,24 +662,169 @@ public bool TryGetValue(TKey key, out TValue value)
             int i = FindEntry(key);
             if (i >= 0)
             {
-                value = entries[i].value;
+                value = _entries[i].value;
                 return true;
             }
             value = default(TValue);
             return false;
         }
         public bool TryAdd(TKey key, TValue value) => TryInsert(key, value, InsertionBehavior.None);
 
-        bool ICollection<KeyValuePair<TKey, TValue>>.IsReadOnly
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool KeyEquals(TKey key0, TKey key1)
         {
-            get { return false; }
+            IEqualityComparer<TKey> customComparer = _customComparer;
+            if (customComparer == null)
+            {
+                if (default(TKey) == null)
+                {
+                    EqualityComparer<TKey> defaultComparer = _defaultComparer;
+                    return defaultComparer.Equals(key0, key1);
+                }
+                else
+                {
+                    if (typeof(TKey) == typeof(byte))
```

Comment:
I do not think that we want to use this pattern and modify every collection in CoreFX to use it. I think it is going in a wrong direction and only fixes the problem for primitive types.

Instead, the JIT should recognize `EqualityComparer<TKey>.Default` as intrinsic - https://github.com/dotnet/coreclr/issues/6688 is about this. So this would be just: EqualityComparer<TKey>.Default.Equals(key0, key1).

Reply:
Added `#if EQUALITYCOMPARER_INTRINSIC`, that work?

Reply:
> only fixes the problem for primitive types

Would be good if C# could express it for all types and all functions, not just relying on the Jit pick up on `EqualityComparer<TKey>.Default.Equals` only  https://github.com/dotnet/csharplang/issues/905

Reply:
@jkotas can you elaborate on what you have in mind for #6688? The source link in that issue seems not to work anymore.

Are you suggesting that `IEqualityComparer<TKey>`'s methods be marked as intrinsics, and then the jit can inject tests at call sites via this interface when `TKey` is known to see if the value is the `EqualityComparer<TKey>.Default`?

Reply:
This this is the file the orignal link pointed to [EqualityComparerHelpers](https://github.com/dotnet/corert/blob/cb4c34eabe549d999d9af2439578606bb38e87b8/src/System.Private.CoreLib/src/Internal/IntrinsicSupport/EqualityComparerHelpers.cs#L144-L168) looks like some constrained static methods it would swap with?

Reply:
> Are you suggesting that IEqualityComparer<TKey>'s methods be marked as intrinsics

Yes.

> the jit can inject tests at call sites via this interface when TKey is known to see if the value is the EqualityComparer<TKey>.Default

I meant that the JIT should recognize `EqualityComparer<TKey>.Default.Equals(a,b)` pattern and replace it with the expanded code. For example, the following https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/ValueTuple.cs#L740 should be just comparing integers directly if I have tuple of integers. It does the virtual calls today.

If it is too hard to recognize `EqualityComparer<TKey>.Default.Equals(a,b)` pattern in the JIT, I would be ok to add a static `EqualityComparer<T>.Equals(a,b)` method and recognize that as intrinsic. It would need to be approved as public API to be widely usable and all relevant places changed to call it, so it is less preferred.